File: RandomPlayer/RPSwipableTVCell.m
//
//  RPSwipableTVCell.m
//  RandomPlayer
//
//  Created by Jonathan Duss on 31.05.14.
//  Copyright (c) 2014 Jonathan Duss. All rights reserved.
//

#import "RPSwipableTVCell.h"
#import "RPCellViewBehind.h"


@interface RPSwipableTVCell()
@property (nonatomic, strong) RPCellViewBehind *rightView;
@property (nonatomic, weak) UIView *leftView;
@property (nonatomic, strong) UISwipeGestureRecognizer *reco;
@property NSTimeInterval lastMovement;
@property float touchSeparationOffset;

@end


@implementation RPSwipableTVCell


- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
    }
    return self;
}



-(void)layoutSubviews
{
    if(_rightView == nil)
    {
        _rightView = [[[NSBundle mainBundle] loadNibNamed:@"cellViewBehind" owner:self options:nil] lastObject];
        [_rightView setHidden:YES];
        _leftView = [self.contentView.subviews objectAtIndex:0];
        
        //add action
        [_rightView.buttonCenterLeft addTarget:self action:@selector(buttonCenterLeftPressed) forControlEvents:UIControlEventTouchUpInside];
        [_rightView.buttonCenterRight addTarget:self action:@selector(buttonCenterRightPressed) forControlEvents:UIControlEventTouchUpInside];
        [_rightView.buttonLeft addTarget:self action:@selector(buttonLeft) forControlEvents:UIControlEventTouchUpInside];
        [_rightView.buttonRight addTarget:self action:@selector(buttonRight) forControlEvents:UIControlEventTouchUpInside];
        
        //resize and add behindView to the cell
        [self.contentView insertSubview:_rightView belowSubview:_leftView];
        CGRect frame = self.contentView.frame;
        _rightView.frame = CGRectMake(frame.size.width - _rightViewOffSet, frame.origin.y, frame.size.width, frame.size.height);
        
        
        UIPanGestureRecognizer *pan = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(paning:)];
        pan.delegate = self;
        
        [self.contentView addGestureRecognizer:pan];
    }
}

- (void)setSelected:(BOOL)selected animated:(BOOL)animated
{
    [super setSelected:selected animated:animated];
    
    // Configure the view for the selected state
}


-(void)prepareForReuse
{
    //Reset the cell by setting defaults position for view
    [_rightView setHidden:YES];
    UIView *contentView = self.contentView;
    CGRect newFVFrame = CGRectMake( 0,
                                   contentView.frame.origin.y,
                                   contentView.frame.size.width,
                                   contentView.frame.size.height);
    
    CGRect newBVFrame = CGRectMake(contentView.frame.size.width - _rightViewOffSet,
                                   contentView.frame.origin.y,
                                   contentView.frame.size.width,
                                   contentView.frame.size.height);
    _rightView.frame = newBVFrame;
    _leftView.frame = newFVFrame;
    
}


#pragma mark - Gesture Recognizer stuff
//############################################################################
/*! Set if the pan gesture should begin or not. If it detects a swipe, it begins */
- (BOOL)gestureRecognizerShouldBegin:(UIGestureRecognizer *)gestureRecognizer
{
    if([gestureRecognizer isMemberOfClass:[UIPanGestureRecognizer class]]){
        UIPanGestureRecognizer *pan = (UIPanGestureRecognizer *)gestureRecognizer;
        CGPoint rec = [pan velocityInView:self.contentView];
        //DLog(@"Velocity: %f, %f", rec.x, rec.y);
        
        if(abs(rec.y) / abs(rec.x) > 1/3){
            return false;
        }
    }
    else
    {
        if([_rightView pointInside:[gestureRecognizer locationInView:[self.contentView.window.subviews objectAtIndex:0]] withEvent:UIEventTypeTouches] ){
            DLog(@"hello");
            return false;
        }
    }
    return true;
}


-(BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer
{
    return NO;
}



//############################################################################
/*! Handle pan gesture */
-(void)paning:(UIPanGestureRecognizer *)pan
{
    if(pan.state == UIGestureRecognizerStateBegan)
    {
        //Correct to do like if the finger was on the separation
        //If finger is 50pixel left from separation, and swipe, then the position of
        // the separation will be position of finger + 50px. OffsetWill be +50.
        
        //Need to add leftViewOffSet  as the separation is in fact where the view begin
        // and thus it can be slighly behind the leftView. We add it, so we can deal like
        //if it was not the case.
        //If first button (80px) not visible. leftViewOffset will be 80px. So
        //The position of the separation appearing on screen will be: position of
        //rightView + 80.
        float positionSeparation = _rightView.frame.origin.x + _rightViewOffSet;
        float positionTouch = [pan locationInView:self.contentView].x;
        
        _touchSeparationOffset = positionSeparation - positionTouch;
        
        DLog(@"\nx: %f     toucheseparation offSet %f",positionTouch, _touchSeparationOffset);
        
        [_rightView setHidden:NO];
    }
    else if (pan.state == UIGestureRecognizerStateChanged)
    {
        [self panMoved:pan];
    }
    else if(pan.state == UIGestureRecognizerStateEnded){
        [self panEnded:pan];
    }
}


//############################################################################
/*!User pan the cell. Updating the position to animate*/
-(void)panMoved:(UIGestureRecognizer *)pan
{
    DLog(@"heigh button: %f", _rightView.buttonCenterLeft.frame.size.height);
    DLog(@"contentView height: %f", self.contentView.frame.size.height);
    DLog(@"self heigh %f",self.frame.size.height);
    
    //Where the user touch
    float x = [pan locationInView:self.contentView].x;
    //give position of finger relative to separation. Negative = finger on left (wrt sep.)
    
    //Correct to do like if the finger was on the separation
    //If finger is 50pixel left from separation, and swipe, then the position of
    // the separation will be position of finger + 50px
    float touchPositionCorrected = x + _touchSeparationOffset ;
    
    //In case the user swipe on the wrong direction, the swipe decrease
//    if(touchPositionCorrected < _frontViewOffSet ){
//        touchPositionCorrected = _frontViewOffSet;
//    }
//    else if (touchPositionCorrected > self.contentView.frame.size.width){
//        touchPositionCorrected = self.contentView.frame.size.width;
//    }
    
    UIView *contentView = self.contentView;
    
    CGRect newFVFrame = CGRectMake( - contentView.frame.size.width + touchPositionCorrected, // ERREUR LA !!!!!!
                                   contentView.frame.origin.y,
                                   contentView.frame.size.width,
                                   contentView.frame.size.height);
    
    CGRect newBVFrame = CGRectMake(touchPositionCorrected - _rightViewOffSet,
                                   contentView.frame.origin.y,
                                   contentView.frame.size.width,
                                   contentView.frame.size.height);
    
    _rightView.frame = newBVFrame;
    _leftView.frame = newFVFrame;
}

//############################################################################
/*!Pan Ended: animate the pan*/
-(void)panEnded:(UIPanGestureRecognizer *)pan
{
    CGRect frame = _rightView.frame;
    float xPosition = frame.origin.x;
    
    UIView *contentView = self.contentView;
    
    double direction = [pan velocityInView:self.contentView].x;
    
    
    
    if(direction > 0){
        //direction > 0 mean last movement was to right
        double velocity = [pan velocityInView:contentView].x;
        
        NSTimeInterval neededTime = (contentView.frame.size.width - xPosition) / velocity * 1.1; //1.1 because of curve slowing
        
        DLog(@"%f",_rightViewOffSet);
        
        if(neededTime < MAX_CELL_ANIMATION_DURATION){
            
            [UIView animateWithDuration:neededTime delay:0 options:UIViewAnimationOptionCurveEaseInOut animations:^(void){
                CGRect newFVFrame = CGRectMake( 0,
                                               contentView.frame.origin.y,
                                               contentView.frame.size.width,
                                               contentView.frame.size.height);
                
                CGRect newBVFrame = CGRectMake(contentView.frame.size.width - _rightViewOffSet,
                                               contentView.frame.origin.y,
                                               contentView.frame.size.width,
                                               contentView.frame.size.height);
                _rightView.frame = newBVFrame;
                _leftView.frame = newFVFrame;
            }completion:nil];
        }
        else
        {
            [UIView animateWithDuration:MAX_CELL_ANIMATION_DURATION delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^(void){
                CGRect newFVFrame = CGRectMake( 0,
                                               contentView.frame.origin.y,
                                               contentView.frame.size.width,
                                               contentView.frame.size.height);
                
                CGRect newBVFrame = CGRectMake(contentView.frame.size.width - _rightViewOffSet,
                                               contentView.frame.origin.y,
                                               contentView.frame.size.width,
                                               contentView.frame.size.height);
                _rightView.frame = newBVFrame;
                _leftView.frame = newFVFrame;
            }completion:nil];
        }
    }
    else {
        //Swipe to the left
        double velocity = abs([pan velocityInView:contentView].x);
        
        
        NSTimeInterval neededTime = xPosition / velocity * 1.1; //1.1 because of curve slowing
        
        if(neededTime < MAX_CELL_ANIMATION_DURATION){
            
            [UIView animateWithDuration:neededTime delay:0 options:UIViewAnimationOptionCurveEaseInOut animations:^(void){
                CGRect newFVFrame = CGRectMake(- contentView.frame.size.width + _rightViewOffSet,
                                               contentView.frame.origin.y,
                                               contentView.frame.size.width,
                                               contentView.frame.size.height);
                
                CGRect newBVFrame = CGRectMake(0,
                                               contentView.frame.origin.y,
                                               contentView.frame.size.width,
                                               contentView.frame.size.height);
                _rightView.frame = newBVFrame;
                _leftView.frame = newFVFrame;
            }completion:nil];
        }
        else
        {
            [UIView animateWithDuration:MAX_CELL_ANIMATION_DURATION delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^(void){
                CGRect newFVFrame = CGRectMake( - contentView.frame.size.width + _rightViewOffSet,
                                               contentView.frame.origin.y,
                                               contentView.frame.size.width,
                                               contentView.frame.size.height);
                
                CGRect newBVFrame = CGRectMake(0,
                                               contentView.frame.origin.y,
                                               contentView.frame.size.width,
                                               contentView.frame.size.height);
                _rightView.frame = newBVFrame;
                _leftView.frame = newFVFrame;
            }completion:nil];
        }
    }
}



-(void)hideBehindCell
{
    UIView *contentView = self.contentView;
    
    [UIView animateWithDuration:MAX_CELL_ANIMATION_DURATION delay:0 options:UIViewAnimationOptionCurveEaseOut animations:^(void){
        CGRect newFVFrame = CGRectMake( 0,
                                       contentView.frame.origin.y,
                                       contentView.frame.size.width,
                                       contentView.frame.size.height);
        
        CGRect newBVFrame = CGRectMake(contentView.frame.size.width - _rightViewOffSet,
                                       contentView.frame.origin.y,
                                       contentView.frame.size.width,
                                       contentView.frame.size.height);
        _rightView.frame = newBVFrame;
        _leftView.frame = newFVFrame;
    }completion:nil];
    
}


#pragma mark - Call the delegate

//These methods are used to set action on the buttons "behind" the cell

-(void)buttonLeftPressed
{
    if([_delegate respondsToSelector:@selector(buttonLeftPressed:)]){
        [self.delegate buttonLeftPressed:self];
    }
    else
    {
        WLog(@"WARNING: %@ does not respond to buttonLeftPressed", _delegate)
    }
}

-(void)buttonCenterLeftPressed
{
    
    if([_delegate respondsToSelector:@selector(buttonLeftPressed:)]){
        [self.delegate buttonCenterLeftPressed:self];
    }
    else
    {
        WLog(@"WARNING: %@ does not respond to buttonCenterLeftPressed", _delegate)
    }
}

-(void)buttonCenterRightPressed
{
    if([_delegate respondsToSelector:@selector(buttonLeftPressed:)]){
        [self.delegate buttonCenterRightPressed:self];
    }
    else
    {
        WLog(@"WARNING: %@ does not respond to buttonCenterRightPressed", _delegate)
    }
}

-(void)buttonRightPressed
{
    if([_delegate respondsToSelector:@selector(buttonLeftPressed:)]){
        [self.delegate buttonRightPressed:self];
    }
    else
    {
        WLog(@"WARNING: %@ does not respond to buttonRightPressed", _delegate)
    }
}

@end

File: RandomPlayer/RPSwipableTVCell.swift
//
//  RPSwipableTVCell.swift
//  RandomPlayer
//
//  Created by Jonathan Duss on 12.08.14.
//  Copyright (c) 2014 Jonathan Duss. All rights reserved.
//

import UIKit





protocol RPSwipableTVCellDelegate {
    func buttonLeftPressed(_ cell : RPSwipableTVCell!)
    func buttonCenterLeftPressed(_ cell : RPSwipableTVCell!)
    func buttonRightPressed(_ cell : RPSwipableTVCell!)
    func buttonCenterRightPressed(_ cell : RPSwipableTVCell!)
}




class RPSwipableTVCell: UITableViewCell {
    
    let MAX_CELL_ANIMATION_DURATION = 0.5
    
    var behindView : RPCellViewBehind?
    
    var behindViewOffSet : CGFloat = 0.0
    
    var frontView : UIView?
    
    var lastMovement = TimeInterval(0)
    var touchOffSet = 0.0
    
    var delegate : RPSwipableTVCellDelegate?
    
    
    
//    override init() {
//        super.init()
//    }
//    
//    override     
//    init(frame: CGRect) {
//        super.init(frame: frame)
//    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override init(style: UITableViewCellStyle, reuseIdentifier: String!) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        // Initialization code
        
        behindView =  Bundle.main.loadNibNamed("cellViewBehind", owner: self, options: nil)?.last as? RPCellViewBehind
        frontView = self.contentView.subviews[0] as? UIView
        behindView?.isHidden = true
        
        self.contentView.insertSubview(behindView!, belowSubview: frontView!)
        
        

        //add actions on buttons
        //TODO
        behindView?.buttonLeft?.addTarget(self, action: #selector(RPSwipableTVCell.buttonLeftPressed), for: UIControlEvents.touchUpInside)
        behindView?.buttonCenterLeft?.addTarget(self, action: #selector(RPSwipableTVCell.buttonCenterLeftPressed), for: UIControlEvents.touchUpInside)
        behindView?.buttonCenterRight?.addTarget(self, action: #selector(RPSwipableTVCell.buttonCenterRightPressed), for: UIControlEvents.touchUpInside)
        behindView?.buttonRight?.addTarget(self, action: #selector(RPSwipableTVCell.buttonRightPressed), for: UIControlEvents.touchUpInside)
    }
    
    
    override func prepareForReuse() {
        //super.prepareForReuse()
        frontView?.frame = self.contentView.frame
        behindView?.isHidden = true
    }
    
    
    
    override func layoutSubviews() {
 
        //do layout
        if(frontView != nil && behindView != nil){
        
            //find offset
            if let b = behindView?.buttonLeft {
                behindViewOffSet = b.frame.origin.x
            }
            else if let b = behindView?.buttonCenterLeft {
                behindViewOffSet = b.frame.origin.x
            } else if let b = behindView?.buttonCenterRight {
                behindViewOffSet = b.frame.origin.x
            } else if let b = behindView?.buttonRight {
                behindViewOffSet = b.frame.origin.x
            }
            
            if let bv = behindView {
                if(bv.isHidden){
                    bv.frame = self.contentView.frame
                    frontView?.frame = self.contentView.frame
                }
                else {
                    let newFrame = CGRect(
                        x: -frame.size.width + self.behindViewOffSet,
                        y: contentView.frame.origin.y,
                        width: contentView.frame.size.width,
                        height: contentView.frame.size.height)
                    bv.frame = self.contentView.frame
                    frontView?.frame = newFrame
                }
            }
            
            
            let pan = UIPanGestureRecognizer(target: self, action: #selector(RPSwipableTVCell.paning(_:)))
            pan.delegate = self
            
            self.addGestureRecognizer(pan)
        }
        else {
            NSException(name: NSExceptionName(rawValue: "Error RPSwipableTVCell"), reason: "Can't find enough views. Check that you have in the custom cell: a view over the content view. You have to add UI element in that view", userInfo: nil).raise()
        }
        
        super.layoutSubviews()
    }
    
    
    //############################################################################
    //############################################################################
    //MARK : Gesture stuff
    
    override func gestureRecognizerShouldBegin(_ gestureRecognizer: UIGestureRecognizer) -> Bool {
        if let pan = gestureRecognizer as? UIPanGestureRecognizer {
//        if(gestureRecognizer.isMember(of: UIPanGestureRecognizer)){
        
            //Pan is swiping the cell.
            //only activate it if the velocity of the pan is above a threshold
            
            let velocity = pan.velocity(in: self.contentView)
            
            if(abs(velocity.y) / abs(velocity.x) > 1/4){
                return false
            }
        }
        else {
            if let view = behindView {
                let viewInside = (self.contentView.window!).subviews[0] 
                //var eventToTest = UIEvent()
                
                if(view.point(inside: gestureRecognizer.location(in: viewInside), with: nil)){
                    return false
                }
            }

        }

        return true
    }
    
    
    
    override func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer)-> Bool {
        return false
    }
    
    func paning(_ pan : UIPanGestureRecognizer) {
        
        if(pan.state == UIGestureRecognizerState.began){
            behindView?.isHidden = false
            if let fv = frontView {
                let frontViewPositionX = fv.frame.origin.x
                let touchPositionX = pan.location(in: self.contentView).x
                
                touchOffSet = Double(frontViewPositionX) - Double(touchPositionX)
                behindView?.isHidden = false
            }
        }
        else if (pan.state == UIGestureRecognizerState.changed) {
            panMoved(pan)
        }
        else if(pan.state == UIGestureRecognizerState.ended) {
            panEnded(pan)
        }
    }
    
    
    /**The cell is being swiped currently*/
    func panMoved(_ pan : UIGestureRecognizer) {
        
        self.selectionStyle = UITableViewCellSelectionStyle.none

        
        let touchPositionX = pan.location(in: self.contentView).x
        var touchPositionCorrectedX = touchPositionX + CGFloat(touchOffSet)
        
        if(touchPositionCorrectedX > 0){
            touchPositionCorrectedX = log2(touchPositionCorrectedX)
        }
        else if(touchPositionCorrectedX < (-1 * contentView.frame.size.width + behindViewOffSet)) {
            touchPositionCorrectedX = (-1 * contentView.frame.size.width + behindViewOffSet) - log2(abs(touchPositionCorrectedX + contentView.frame.size.width - behindViewOffSet))
        }
        
        
        let newFrame = CGRect(x: touchPositionCorrectedX,
            y: contentView.frame.origin.y,
            width: contentView.frame.width,
            height: contentView.frame.size.height)
        
        frontView?.frame = newFrame
        
    }
    
    
    /**The cell swiping has ended*/
    func panEnded(_ pan : UIPanGestureRecognizer) {
        
        if let fv = frontView{
            let speedUpConstant = CGFloat(1.1)
            
            let frontViewXPosition = fv.frame.origin.x
            
            let velocity = pan.velocity(in: self.contentView).x
            
            if(velocity > 0) {
                //velocity < 0 -> last movement went to the right (cell become closed)
                
                var neededTime = TimeInterval((contentView.frame.size.width - frontViewXPosition) / velocity * speedUpConstant)
                
                if(neededTime > MAX_CELL_ANIMATION_DURATION){
                    neededTime = MAX_CELL_ANIMATION_DURATION
                }

                
                //animated the cell so it continue to be "swiped" and hide completely what is behind
                UIView.animate(withDuration: neededTime, delay: 0, options: UIViewAnimationOptions.curveEaseOut,
                    animations: {() -> Void in
                        fv.frame = self.contentView.frame
                    },
                    completion:{(c : Bool) -> Void in
                        if(c){
                            self.behindView?.isHidden = true
                        }
                })
                
                

            }
            else {
                //velocity > 0 -> last movement went to the left
                
                var neededTime = (contentView.frame.size.width - frontViewXPosition) / velocity * speedUpConstant
                
                if(neededTime > CGFloat(MAX_CELL_ANIMATION_DURATION)){
                    neededTime = CGFloat(MAX_CELL_ANIMATION_DURATION)
                }
                
                //animated the cell so it continue to be "swiped" and show completely what is behind
                UIView.animate(withDuration: TimeInterval(neededTime), delay: 0, options: UIViewAnimationOptions.curveEaseOut, animations: {() -> Void in
                    let frame = self.contentView.frame
                    
                    let newFrame = CGRect(
                        x: -frame.size.width + self.behindViewOffSet,
                        y: frame.origin.y,
                        width: frame.size.width,
                        height: frame.size.height)
                    fv.frame = newFrame
                    }, completion: nil)
            }
        }
    }

    //############################################################################################
    //############################################################################################
    
    //UITableViewCell stuff
    override func setSelected(_ selected: Bool, animated: Bool) {
        super.setSelected(selected, animated: animated)
        
        // Configure the view for the selected state
    }
    
    
    /**Show the view "behind the cell". (Show the buttons) */
    func hideBehindCell(){
        
        if let fv = frontView{
            let neededTime = MAX_CELL_ANIMATION_DURATION
            
            UIView.animate(withDuration: neededTime, delay: 0, options: UIViewAnimationOptions.curveEaseOut,
                animations: {() -> Void in
                    fv.frame = self.contentView.frame
                },
                completion:{(c : Bool) -> Void in
                    if(c){
                        self.behindView?.isHidden = true
                    }
            })
            
        }
        
    }
    
    //############################################################################################
    //############################################################################################
    //Handle buttons
    func buttonLeftPressed(){
        self.delegate?.buttonCenterLeftPressed(self)
    }
    
    func buttonCenterLeftPressed(){
        delegate?.buttonCenterLeftPressed(self)
    }
    
    func buttonRightPressed(){
        delegate?.buttonRightPressed(self)
    }
    
    func buttonCenterRightPressed(){
        delegate?.buttonCenterRightPressed(self)
    }
    
}
